src/VisualStudio/IntegrationTests/CSharp/CSharpAutomaticBraceCompletion.cs

```diff
@@ -0,0 +1,146 @@
+            _visualStudio.Dispose();
+        }
+
+        [Fact]
+        public void BracesInsertionAndTabCompleting()
+        {
+            _editorWindow.Text = @"class C {
+    void Foo() {
+        // Marker
+    }
+}";
+
+            _editorWindow.PlaceCursor("// Marker");
+
+            _editorWindow.TypeText("if (true) {");
+            VerifyEditorTextBeforeAndAfterCursor("        if (true) { ", "}");
+
+            _editorWindow.TypeText("\t");
+            VerifyEditorTextBeforeCursor("        if (true) { }");
+        }
+
+        [Fact]
+        public void BracesOvertyping()
+        {
+            _editorWindow.Text = @"class C {
+    void Foo() {
+        // Marker
+    }
+}";
+
+            _editorWindow.PlaceCursor("// Marker");
+
+            _editorWindow.TypeText("if (true) {");
+
+            _editorWindow.TypeText("}");
+            VerifyEditorTextBeforeCursor("        if (true) { }");
+        }
+
+        [Fact]
+        public void BracesOnReturnNoFormattingOnlyIndentationBeforeCloseBrace()
+        {
+            _editorWindow.Text = @"class C {
+    void Foo() {
+        // Marker
+    }
+}";
+
+            _editorWindow.PlaceCursor("// Marker");
+
+            _editorWindow.TypeText("if (true) {");
+
+            _editorWindow.TypeText("\rvar a = 1;");
+            VerifyEditorTextBeforeCursor("            var a = 1;");
+        }
+
+        [Fact]
+        public void BracesOnReturnOvertypingTheClosingBrace()
+        {
+            _editorWindow.Text = @"class C {
+    void Foo() {
+        // Marker
+    }
+}";
+
+            _editorWindow.PlaceCursor("// Marker");
+
+            _editorWindow.TypeText("if (true) {");
+
+            _editorWindow.TypeText("\rvar a = 1;}");
+            VerifyEditorTextBeforeCursor("        }");
+
+            VerifyEditorTextContains(@"if (true)
+        {
+            var a = 1;
+        }
+");
+        }
+
+        [Fact]
+        public void BracesOnReturnWithNonWhitespaceSpanInside() // DevDiv 653540
+        {
+            _editorWindow.Text = string.Empty;
+            _editorWindow.TypeText("class A { int i;\r");
+
+            VerifyEditorTextAfterCursor("}");
+            VerifyEditorTextContains(@"class A { int i;
+}");
+        }
+
+        private void VerifyEditorTextAfterCursor(string expectedText) => VerifyEditorTextBeforeAndAfterCursor(string.Empty, expectedText);
+
+        private void VerifyEditorTextBeforeCursor(string expectedText) => VerifyEditorTextBeforeAndAfterCursor(expectedText, string.Empty);
+
+        private void VerifyEditorTextContains(string expectedText)
+        {
+            Assert.Contains(expectedText, _editorWindow.Text);
+        }
+
+        private void VerifyEditorTextBeforeAndAfterCursor(string expectedTextBeforeCursor, string expectedTextAfterCursor)
+        {
+            if (expectedTextBeforeCursor != null)
+            {
+                Assert.Equal(expectedTextBeforeCursor, _editorWindow.CurrentLineTextBeforeCursor);
```

Comment:
Ditto; let's inline the assertions into the test versus hiding them behind a few layers of indirection.